crawl-ref/source/newgame.cc

```diff
@@ -182,38 +182,34 @@ static void _resolve_species(newgame_def& ng, const newgame_def& ng_choice)
         return;
 
     case SP_VIABLE:
-        // If we have already picked a job, find a species that recommends it
-        if (is_starting_job(ng.job))
-        {
-            auto species_recommended_by_job =
-                filter_enum(species_type, [&](species_type species) {
-                    return job_recommends_species(ng.job, species);
-                });
-            // Note: we know the array has at least one element because all
-            // starting jobs must recommend at least one opposite species.
-            ng.species = species_recommended_by_job[
-                random2(species_recommended_by_job.size())];
-        }
-        else
-            ng.species = random_species();
-        return;
-
     case SP_RANDOM:
-        // If we have already picked a job, find a species that allows it
+    {
+        species_type[] candidate_species;
         if (is_starting_job(ng.job))
         {
-            auto species_allowed_by_job =
-                filter_enum(species_type, [&](species_type species) {
-                    return character_is_allowed(species, ng.job);
-                });
-            if (species_allowed_by_job.size() == 0)
-                die("Selected job allows no species?!");
-            ng.species = species_allowed_by_job[
-                random2(species_allowed_by_job.size())];
+            switch (ng_choice.species)
+            {
+            case SP_VIABLE:
+                // Note: we know the array has at least one element because all
+                // starting jobs must recommend at least one species.
+                candidate_species =
+                    filter_enum(species_type, [&](species_type species) {
+                        return job_recommends_species(ng.job, species);
+                    });
+            case SP_RANDOM:
+                candidate_species =
+                    filter_enum(species_type, [&](species_type species) {
+                        return character_is_allowed(species, ng.job);
+                    });
+                if (candidate_species.size() == 0)
```

Comment:
this test should be outside the inner switch, so it tests for both VIABLE and RANDOM cases